### PR TITLE
DB-10882 Support updatable cursors

### PR DESCRIPTION
**Pull request by hatyo**

### CODE REVIEW

File: splice_machine/src/main/java/com/splicemachine/derby/stream/function/ScrollInsensitiveFunction.java [hunk 1]
```diff
@@ -41,6 +41,9 @@ public ExecRow call(ExecRow execRow) throws Exception {
             this.operationContext.recordRead();
             op.setCurrentRow(execRow);
             this.operationContext.recordProduced();
+            if(!op.isForUpdate()) {
```
Comment (hatyo) [inline splice_machine/src/main/java/com/splicemachine/derby/stream/function/ScrollInsensitiveFunction.java:44]: @dgomezferro this prevents serializing rowId. Should we added to other functions that could be pushed down to Spark executors and used by the `ResultStreamer`?

File: db-engine/src/main/java/com/splicemachine/db/impl/sql/compile/DeleteNode.java [hunk 2]
```diff
@@ -325,19 +325,11 @@ else if (cursorTargetTableName != null)
...
-                columnReference.setSource(rowIdColumn);
-                columnReference.setNestingLevel(targetTable.getLevel());
-                columnReference.setSourceLevel(targetTable.getLevel());
                 rowLocationColumn =
                         (ResultColumn) getNodeFactory().getNode(
                                 C_NodeTypes.RESULT_COLUMN,
                                 COLUMNNAME,
-                                columnReference,
+                                rowIdColumn,
                                 getContextManager());
```
Comment (dgomezferro) [inline db-engine/src/main/java/com/splicemachine/db/impl/sql/compile/DeleteNode.java:333]: Why is this change needed? @ascend1 @msirek can you comment on whether this could cause issues?

File: splice_machine/src/main/java/com/splicemachine/derby/stream/function/ScrollInsensitiveFunction.java (hunk 1 above)
Comment (dgomezferro) [reply splice_machine/src/main/java/com/splicemachine/derby/stream/function/ScrollInsensitiveFunction.java:44]: I think this is the only case, at least for now

File: splice_machine/src/test/java/com/splicemachine/derby/test/framework/SpliceIndexWatcher.java [hunk 3]
```diff
@@ -31,7 +31,7 @@
     protected String tableName;
     protected String tableSchemaName;
     protected String createString;
-    protected String indexName;
+    public final String indexName;
```
Comment (dgomezferro) [inline splice_machine/src/test/java/com/splicemachine/derby/test/framework/SpliceIndexWatcher.java:34]: Nitpick, a getter for this would be better

**Review by dgomezferro — CHANGES_REQUESTED** [review]
Great job @hatyo , that was so fast! I only have some requests:
1. Can you add some more tests covering triggers (update/insert/delete) and concurrent updates? I've tested update triggers and they seem to work well, I'm interested in preventing regressions. Concurrent updates also seem to work well, if a row is deleted concurrently a WW conflict is raised for the update. After that the same cursor position can't be updated again, not sure if that's expected.
2. I think this is broken if we force Spark execution on the connection string : connect '...;useSpark=true' I get `ERROR 42X30: Cursor 'SQL_CURLH000C1' not found. Verify that autocommit is OFF.`

**Comment by hatyo** [comment]
Thanks @dgomezferro for the comments, I didn't know we could enforce execution on Spark in the JDBC connection string, I will look into the failure and address the other comments.

**Comment by hatyo** [comment]
jenkins please test branch @cdh6.3.0,skipTestsLongerThan2Minutes

**Comment by spliceci** [comment]
TEST FAILED
http://nexus.splicemachine.com:8080/job/spliceengine-PR/job/Splice-PR-Branch/2167/

File: db-engine/src/main/java/com/splicemachine/db/impl/sql/compile/DeleteNode.java (hunk 2 above)
Comment (hatyo) [reply db-engine/src/main/java/com/splicemachine/db/impl/sql/compile/DeleteNode.java:333]: If we don't expose the `CURRENT_ROW_LOCATION_NODE` in the RCL for update/delete nodes we will not generate the corresponding `getRowLocation` in the projection expression, which means we will not be able to retrieve the rowId from the projection node.
For more information please have a look at `com.splicemachine.db.impl.sql.compile.ResultColumnList#generateEvaluatedRow`

File: splice_machine/src/test/java/com/splicemachine/derby/test/framework/SpliceIndexWatcher.java (hunk 3 above)
Comment (hatyo) [reply splice_machine/src/test/java/com/splicemachine/derby/test/framework/SpliceIndexWatcher.java:34]: fixed it.

File: splice_machine/src/main/java/com/splicemachine/derby/stream/function/ScrollInsensitiveFunction.java (hunk 1 above)
Comment (hatyo) [reply splice_machine/src/main/java/com/splicemachine/derby/stream/function/ScrollInsensitiveFunction.java:44]: ok

**Comment by hatyo** [comment]
I found some bugs in the implementation and fixed them. I also added some tests for interaction with triggers and concurrent transactions updating PK and non-PK columns.

**Review by dgomezferro — APPROVED** [review]

**Comment by spliceci** [comment]
TEST FAILED
http://nexus.splicemachine.com:8080/job/spliceengine-PR/job/Splice-PR-Branch/2173/

**Comment by hatyo** [comment]
jenkins please test branch @cdh6.3.0,skipTestsLongerThan2Minutes

**Comment by spliceci** [comment]
TEST FAILED
http://nexus.splicemachine.com:8080/job/spliceengine-PR/job/Splice-PR-Branch/2174/

**Review by msirek — APPROVED** [review]

**Comment by hatyo** [comment]
> TEST FAILED
> [http://nexus.splicemachine.com:8080/job/spliceengine-PR/job/Splice-PR-Branch/2174](http://nexus.splicemachine.com:8080/job/spliceengine-PR/job/Splice-PR-Branch/2174/)

Due to DB-9786.

**Comment by hatyo** [comment]
jenkins please test branch @cdh6.3.0,skipTestsLongerThan2Minutes

**Comment by spliceci** [comment]
TEST FAILED
http://nexus.splicemachine.com:8080/job/spliceengine-PR/job/Splice-PR-Branch/2178/

**Comment by ipraznik-splice** [comment]
jenkins please test branch @cdh6.3.0,skipTestsLongerThan2Minutes

**Comment by spliceci** [comment]
TEST FAILED
http://nexus.splicemachine.com:8080/job/spliceengine-PR/job/Splice-PR-Branch/2179/

**Comment by hatyo** [comment]
jenkins please test branch @cdh6.3.0,skipTestsLongerThan2Minutes

**Comment by spliceci** [comment]
TEST FAILED
http://nexus.splicemachine.com:8080/job/spliceengine-PR/job/Splice-PR-Branch/2181/

**Review by carolp-503 — APPROVED** [review]
+1

**Comment by hatyo** [comment]
jenkins please test branch @cdh6.3.0,skipTestsLongerThan2Minutes

File: db-engine/src/main/java/com/splicemachine/db/impl/sql/execute/IndexValueRow.java [hunk 4]
```diff
@@ -54,7 +54,7 @@
  */
 public class IndexValueRow implements ExecIndexRow, Serializable {
 
-	private ExecRow valueRow;
+	private ExecRow                                     valueRow;
```
Comment (dgomezferro) [inline db-engine/src/main/java/com/splicemachine/db/impl/sql/execute/IndexValueRow.java:57]: Remove extra spacing

File: splice_machine/src/main/java/com/splicemachine/derby/stream/function/SetCurrentLocatedRowAndRowKeyFunction.java [hunk 5]
```diff
@@ -17,27 +17,31 @@
...
     public SetCurrentLocatedRowAndRowKeyFunction(OperationContext operationContext) {
         super(operationContext);
+        locatorFunction = new BaseRowLocatorFunction<>();
     }
 
     @Override
     public ExecRow call(ExecRow locatedRow) throws Exception {
         getOperation().setCurrentRow(locatedRow);
         getOperation().setCurrentRowLocation(new HBaseRowLocation(locatedRow.getKey()));
+        getOperation().setCurrentBaseRowLocation(locatorFunction.call(operationContext));
```
Comment (dgomezferro) [inline splice_machine/src/main/java/com/splicemachine/derby/stream/function/SetCurrentLocatedRowAndRowKeyFunction.java:44]: We are setting this unconditionally, not just for "FOR UPDATE" statements. Is it possible to restrict it?

File: splice_machine/src/main/java/com/splicemachine/derby/stream/function/BaseRowLocatorFunction.java [hunk 6]
```diff
@@ -0,0 +1,55 @@
...
+
+import com.splicemachine.db.iapi.sql.execute.ExecRow;
+import com.splicemachine.derby.iapi.sql.execute.SpliceOperation;
+import com.splicemachine.derby.impl.sql.execute.operations.IndexRowToBaseRowOperation;
+import com.splicemachine.derby.impl.sql.execute.operations.ProjectRestrictOperation;
+import com.splicemachine.derby.impl.sql.execute.operations.ScrollInsensitiveOperation;
+import com.splicemachine.derby.impl.sql.execute.operations.TableScanOperation;
+import com.splicemachine.derby.stream.iapi.OperationContext;
+
+public class BaseRowLocatorFunction <Op extends SpliceOperation> extends SpliceFunction<Op, OperationContext<Op>,ExecRow> {
```
Comment (dgomezferro) [inline splice_machine/src/main/java/com/splicemachine/derby/stream/function/BaseRowLocatorFunction.java:24]: I don't think this needs to be a SpliceFunction, the logic could even be a static function. I'm ok with it being a functional object, but no need to extend SpliceFunction, implement `hasNativeSparkImplementation` etc.

**Review by dgomezferro — CHANGES_REQUESTED** [review]
Nice work! I have some comments, but I like the overall approach. Have you tested Partial Updates using indices in Spark?

File: platform_it/pom.xml [hunk 7]
```diff
@@ -1069,6 +1069,7 @@
                                 <argument>-Dderby.language.updateSystemProcs=false</argument>
                                 <!-- Setting the logStatementText option to true enables logging of all statements. -->
                                 <argument>-Dderby.infolog.append=true</argument>
+                                <argument>-Dderby.debug.true=DumpClassFile</argument>
```
[the remark this answers is not in the document]
Comment (hatyo) [reply platform_it/pom.xml:1076]: done.

File: db-engine/src/main/java/com/splicemachine/db/impl/sql/execute/IndexValueRow.java (hunk 4 above)
Comment (hatyo) [reply db-engine/src/main/java/com/splicemachine/db/impl/sql/execute/IndexValueRow.java:57]: done.

File: splice_machine/src/main/java/com/splicemachine/derby/stream/function/SetCurrentLocatedRowAndRowKeyFunction.java (hunk 5 above)
Comment (hatyo) [reply splice_machine/src/main/java/com/splicemachine/derby/stream/function/SetCurrentLocatedRowAndRowKeyFunction.java:44]: You're right, I restrict it for for-update plans.

File: splice_machine/src/main/java/com/splicemachine/derby/stream/function/BaseRowLocatorFunction.java (hunk 6 above)
Comment (hatyo) [reply splice_machine/src/main/java/com/splicemachine/derby/stream/function/BaseRowLocatorFunction.java:24]: You're right, I wanted initially to make it a full-fledged SpliceFunction that I inject in the plan execution, but along the way I reduced it to a normal static function. I will refactor it accordingly.

**Comment by hatyo** [comment]
> Nice work! I have some comments, but I like the overall approach. Have you tested Partial Updates using indices in Spark?

Thanks @dgomezferro for the comments, I addressed all of them and added a test for index scan with partial update.

**Comment by ipraznik-splice** [comment]
jenkins please test branch @cdh6.3.0,skipTestsLongerThan2Minutes

**Comment by spliceci** [comment]
TEST SUCCEEDED +1

**Comment by spliceci** [comment]
TEST SUCCEEDED +1

File: db-engine/src/main/java/com/splicemachine/db/impl/sql/compile/DeleteNode.java (hunk 2 above)
Comment (dgomezferro) [reply db-engine/src/main/java/com/splicemachine/db/impl/sql/compile/DeleteNode.java:337]: That makes sense, thanks!

**Review by dgomezferro — APPROVED** [review]
Awesome @hatyo , thank you!
